### PR TITLE
feat(FN-1859): refactor upload report logic to SQL

### PR DESCRIPTION
**Pull request by toddym42**
## Introduction :pencil2:
We want to refactor the uploading report logic to use the SQL database.

## Resolution :heavy_check_mark:
- Updates the state machine event logic to upload reports
- Adds a `UtilisationReportEntityMockBuilder` class
- Adds new and updates existing tests

### CODE REVIEW

File: portal-api/src/v1/api.js [hunk 1]
```diff
@@ -249,21 +249,24 @@ const findLatestGefMandatoryCriteria = async () => {
   }
 };
 
-const saveUtilisationReport = async (reportData, reportPeriod, user, fileInfo) => {
+const saveUtilisationReport = async (reportId, reportData, user, fileInfo) => {
```
Comment (JoshBinns2000) [inline portal-api/src/v1/api.js:252]: Needs a JSDoc here

File: portal-api/api-tests/v1/utilisation-report-service/utilisation-report-upload.api-test.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ jest.mock('../../../src/drivers/fileshare', () => ({
 }));
 
 jest.mock('../../../src/v1/api', () => ({
-  saveUtilisationReport: jest.fn().mockResolvedValue({ status: 201, data: { dateUploaded: new Date() } }),
+  saveUtilisationReport: jest.fn().mockResolvedValue({ dateUploaded: new Date() }),
```
Comment (JoshBinns2000) [inline portal-api/api-tests/v1/utilisation-report-service/utilisation-report-upload.api-test.js:21]: This is where the api tests error comes from - we check that the length of `response.data` is equal to 1, but we mock a single object return here. Need to add some docs to make it more clear as to what is returned

File: dtfs-central-api/src/utils/report-period.test.ts [hunk 3]
```diff
@@ -175,4 +177,102 @@ describe('report-period utils', () => {
...
+
+      expect(validationErrors).toEqual([]);
+    });
+
+    const validMonthAndYear: MonthAndYear = {
+      month: 1,
+      year: 2024,
+    };
+
+    it.each`
```
Comment (ChristianMcc) [inline dtfs-central-api/src/utils/report-period.test.ts:202]: Nice tests

File: dtfs-central-api/src/utils/report-period.test.ts [hunk 4]
```diff
@@ -175,4 +177,102 @@ describe('report-period utils', () => {
...
+      ${"'reportPeriod.end.year' is undefined"}           | ${{ start: validMonthAndYear, end: { month: 1, year: undefined } }}    | ${"Report period is not valid: 'reportPeriod.end.year' is required"}
+      ${"'reportPeriod.end.year' is null"}                | ${{ start: validMonthAndYear, end: { month: 1, year: null } }}         | ${"Report period is not valid: 'reportPeriod.end.year' is required"}
+      ${"'reportPeriod.end.year' is less than 2020"}      | ${{ start: validMonthAndYear, end: { month: 1, year: 2019 } }}         | ${"Report period is not valid: 'reportPeriod.end.year' must be between 2020 and 2100"}
+      ${"'reportPeriod.end.year' is greater than 2100"}   | ${{ start: validMonthAndYear, end: { month: 1, year: 2101 } }}         | ${"Report period is not valid: 'reportPeriod.end.year' must be between 2020 and 2100"}
+    `('returns an error when $condition', ({ invalidReportPeriod, errorMessage }) => {
+      // Act
+      const validationErrors = validateReportPeriod(invalidReportPeriod);
+
+      // Assert
+      expect(validationErrors).toEqual(expect.arrayContaining([errorMessage]));
```
Comment (ChristianMcc) [inline dtfs-central-api/src/utils/report-period.test.ts:231]: unnecessary whitespace

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts [hunk 5]
```diff
@@ -8,11 +9,10 @@ import {
 } from '../../../validation/utilisation-report-service/utilisation-report-validator';
 import { UtilisationReportStateMachine } from '../../../../services/state-machines/utilisation-report/utilisation-report.state-machine';
 import { CustomExpressRequest } from '../../../../types/custom-express-request';
-import { AzureFileInfo } from '../../../../types/azure-file-info';
 import { UtilisationReportRawCsvData } from '../../../../types/utilisation-reports';
 import { ApiError } from '../../../../errors';
 
-type PostUploadUtilisationReportRequest2 = CustomExpressRequest<{
+export type PostUploadUtilisationReportRequest2 = CustomExpressRequest<{
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts:15]: What's the need for this type over the other one? Is it purely for the payloadValidator before we cast to the other type after? Reckon we can improve the names if so

File: dtfs-central-api/src/v1/validation/utilisation-report-service/utilisation-report-validator.js [hunk 6]
```diff
@@ -35,7 +35,7 @@ const validateReportId = (reportId) => {
  * @returns {string | null} - Error message or null if valid.
  */
 const validateMonth = (month, propertyName = 'Month') => {
-  if (!month) {
+  if (!month && month !== 0) {
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/validation/utilisation-report-service/utilisation-report-validator.js:38]: mmm javascripty falsey values

File: libs/common/src/sql-db-entities/utilisation-report/utilisation-report.entity.ts [hunk 7]
```diff
@@ -72,8 +74,7 @@ export class UtilisationReportEntity extends AuditableBaseEntity {
     return report;
   }
 
-  // TODO FN-1859 - maybe just update 'this' (and rename back to updateWithUploadDetails) rather than clone and update?
-  public updatedWithUploadDetails({ azureFileInfo, data, uploadedByUserId, requestSource }: UpdateWithUploadDetailsParams): UtilisationReportEntity {
+  public toReportWithUploadDetails({ azureFileInfo, data, uploadedByUserId, requestSource }: UpdateWithUploadDetailsParams): UtilisationReportEntity {
```
Comment (ChristianMcc) [inline libs/common/src/sql-db-entities/utilisation-report/utilisation-report.entity.ts:77]: ditto on naming

File: dtfs-central-api/src/repositories/utilisation-reports-repo/utilisation-report-sql.repo.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ export const UtilisationReportRepo = SqlDbDataSource.getRepository(UtilisationRe
       }),
     );
 
-    const updatedReport = report.updatedWithUploadDetails({
+    const updatedReport = report.toReportWithUploadDetails({
```
[the remark this answers is not in the document]
Comment (ChristianMcc) [reply dtfs-central-api/src/repositories/utilisation-reports-repo/utilisation-report-sql.repo.ts:36]: (i commented ditto where this method is actually defined rather than here!)

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts (hunk 5 above)
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts:15]: Yeh this was something Mike left behind - I believe it just takes everything in `PostUploadUtilisationReportRequest` and sets everything to `unknown`. I'll have a think about a better way of doing this

File: dtfs-central-api/src/v1/validation/utilisation-report-service/utilisation-report-validator.js (hunk 6 above)
Comment (JoshBinns2000) [reply dtfs-central-api/src/v1/validation/utilisation-report-service/utilisation-report-validator.js:38]: Yep - something that would technically work fine, but our error messages seem specific so I wanted our check to be specific too (caught this in one of the unit tests I added)

File: dtfs-central-api/src/utils/report-period.test.ts (hunk 4 above)
Comment (JoshBinns2000) [reply dtfs-central-api/src/utils/report-period.test.ts:231]: Where is the unnecessary whitespace?

File: dtfs-central-api/src/repositories/utilisation-reports-repo/utilisation-report-sql.repo.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoshBinns2000) [reply dtfs-central-api/src/repositories/utilisation-reports-repo/utilisation-report-sql.repo.ts:36]: Discussed: we are going to use the more OOP approach of mutating the state of the actual report rather than copying, updating and returning, meaning the method name is not `updateWithUploadDetails`

File: libs/common/src/sql-db-entities/utilisation-report/utilisation-report.entity.ts (hunk 7 above)
Comment (JoshBinns2000) [reply libs/common/src/sql-db-entities/utilisation-report/utilisation-report.entity.ts:77]: See [here](https://github.com/UK-Export-Finance/dtfs2/pull/2793/files#r1518864638)

File: dtfs-central-api/src/helpers/utilisation-data.helper.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import { UTILISATION_REPORT_HEADERS } from '../constants';
+
+type UtilisationDataCsvRowToSqlEntityParams = {
+  dataEntry: UtilisationReportRawCsvData;
+  requestSource: DbRequestSource;
+};
+
+const asNumberOrDefault = (maybeString: string | undefined, defaultNumber: number) => (maybeString ? Number(maybeString) : defaultNumber);
+
+export const utilisationDataCsvRowToSqlEntity = ({ dataEntry, requestSource }: UtilisationDataCsvRowToSqlEntityParams): UtilisationDataEntity =>
```
Comment (ChristianMcc) [inline dtfs-central-api/src/helpers/utilisation-data.helper.ts:12]: Sorry but going to be a pain and ask for some docs on a bunch of functions, can you just check through the PR and see if there are any functions (not in tests) missing any please

File: dtfs-central-api/src/repositories/utilisation-data-repo/utilisation-data-repo.ts [hunk 10]
```diff
@@ -1,35 +1,13 @@
...
-      monthlyFeesPaidToUkefCurrency: reportDataEntry[UTILISATION_REPORT_HEADERS.FEES_PAID_IN_PERIOD_CURRENCY],
-      paymentCurrency: reportDataEntry[UTILISATION_REPORT_HEADERS.PAYMENT_CURRENCY],
-      paymentExchangeRate: reportDataEntry[UTILISATION_REPORT_HEADERS.PAYMENT_EXCHANGE_RATE]
-        ? Number(reportDataEntry[UTILISATION_REPORT_HEADERS.PAYMENT_EXCHANGE_RATE])
-        : 1,
-      payments: null,
-    }),
-  );
 
+export const getAllUtilisationDataForReport = async ({ _id: reportId, reportPeriod }: UtilisationReport): Promise<UtilisationData[]> => {
```
Comment (ChristianMcc) [inline dtfs-central-api/src/repositories/utilisation-data-repo/utilisation-data-repo.ts:4]: docs

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  fileInfo: AzureFileInfo;
+  reportData: UtilisationReportRawCsvData[];
+  user: { _id: string };
+};
+
+type PreValidationPostUploadUtilisationReportRequest = CustomExpressRequest<{
+  reqBody: Unknown<PostUploadUtilisationReportRequestBody>;
+}>;
+
+export const postUploadUtilisationReportPayloadValidator = (req: PreValidationPostUploadUtilisationReportRequest, res: Response, next: NextFunction) => {
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts:26]: docs

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  return next();
+};
+
+type PostUploadUtilisationReportRequest = CustomExpressRequest<{
+  reqBody: PostUploadUtilisationReportRequestBody;
+}>;
+
+export const postUploadUtilisationReport = async (req: PostUploadUtilisationReportRequest, res: Response) => {
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/post-upload-utilisation-report.controller/index.ts:48]: docs

File: libs/common/src/types/types-helper.ts [hunk 13]
```diff
@@ -7,3 +7,20 @@ export type Prettify<T> = {
...
+ * @example
+ * type Body = {
+ *   id: number;
+ *   name: string;
+ * };
+ *
+ * type PreValidationBody = Unknown<Body>;
+ * //   ^? { id: unknown; name: unknown }
+ */
+export type Unknown<T> = {
```
Comment (ChristianMcc) [inline libs/common/src/types/types-helper.ts:24]: Will this right now only do a shallow copy to unknown e.g. what if I have a type of:

`{
  a: string,
  b: {
    c: string;
    d: string:
  },
}`
? Tbh that sounds like exactly what we want if we don't know the type that's coming into the API in the request :D
Something about the name doesn't feel descriptive enough but the best i've thought of is 'unknownise' which is just worse :D

File: portal-api/src/v1/api-response-types/utilisation-report-response-body.ts [hunk 14]
```diff
@@ -4,8 +4,8 @@ export type UtilisationReportResponseBody = {
   id: number;
   bankId: string;
   status: UtilisationReportReconciliationStatus;
-  uploadedByUserId: string;
+  uploadedByUserId?: string;
```
Comment (ChristianMcc) [inline portal-api/src/v1/api-response-types/utilisation-report-response-body.ts:7]: What is this change for?

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js [hunk 15]
```diff
@@ -100,6 +100,7 @@ const uploadReportAndSendNotification = async (req, res) => {
     const { file } = req;
 
     const { formattedReportPeriod, reportData, reportPeriod, user } = req.body;
+    // TODO FN-1859 - do we need to parse here?
```
Comment (ChristianMcc) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:103]: most likely, I think this is being sent in formData rather than a normal json body of a request because it's being sent with a file. Either way if you want to check this do and then remove comment please :)

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js [hunk 16]
```diff
@@ -108,34 +109,37 @@ const uploadReportAndSendNotification = async (req, res) => {
...
+
+    const existingReport = existingReports[0];
+
+    // If a report has already been uploaded, we should not overwrite it
+    if (existingReport.azureFileInfo) {
       return res.status(409).send('Report for the supplied report period has already been uploaded');
     }
 
-    const fileInfo = await saveFileToAzure(file, parsedUser.bank.id);
+    const fileInfo = await saveFileToAzure(file, bankId);
```
Comment (ChristianMcc) [inline portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:129]: This section here -  are we trying to allow a user to upload a report if it's 'Marked as complete'?
If it's marked as complete it means they shouldnt be able to upload a report for that period. I'd prefer this to actually check that there are no reports that aren't in the 'NOT_RECEIVED' status

**Review by ChristianMcc — COMMENTED** [review]
A couple of nit-picks on documentation sorry - then one comment that I think is on Mike's code that I didn't pick up last time sorry!

File: dtfs-central-api/src/repositories/utilisation-data-repo/utilisation-data-repo.ts (hunk 10 above)
Comment (JoshBinns2000) [reply dtfs-central-api/src/repositories/utilisation-data-repo/utilisation-data-repo.ts:4]: This is a mongo DB repo method that should be removed pretty soon, so is it worth adding a docstring here?

File: libs/common/src/types/types-helper.ts (hunk 13 above)
Comment (JoshBinns2000) [reply libs/common/src/types/types-helper.ts:24]: The name was based on typescript utility types like `Readonly` which just makes all properties `readonly`. Similarly, `Readonly` only does a shallow copy of the type:
```typescript
// lib.es5.d.ts
/**
 * Make all properties in T readonly
 */
type Readonly<T> = {
    readonly [P in keyof T]: T[P];
};
```
The usual pattern I've seen in typescript is that the standard utility type is shallow, and you would extend the type to `DeepReadonly` (for instance) if you want it to be deep.

Naming wise, I agree that `Unknown` isn't clear and I was hoping a better name would come to me 😅. Maybe `WithUnknownProperties`?

File: portal-api/src/v1/api-response-types/utilisation-report-response-body.ts (hunk 14 above)
Comment (JoshBinns2000) [reply portal-api/src/v1/api-response-types/utilisation-report-response-body.ts:7]: `uploadedByUserId` and `dateUploaded` are optional in the entity and so they will also be optional in the response

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js (hunk 15 above)
Comment (JoshBinns2000) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:103]: I think Mike added this, but wanted to leave it in just so we could have a comment left by it on the PR :)

File: portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js (hunk 16 above)
Comment (JoshBinns2000) [reply portal-api/src/v1/controllers/utilisation-report-service/utilisation-report-upload.controller.js:129]: Alright - I will update to check for status instead of whether or not a report has been uploaded

**Review by ChristianMcc — APPROVED** [review]
